airflow/jobs/triggerer_job.py

```diff
@@ -0,0 +1,418 @@
+        try:
+            async for event in trigger.run():
+                self.log.info("Trigger %s fired: %s", self.triggers[trigger_id]['name'], event)
+                self.triggers[trigger_id]["events"] += 1
+                self.events.append((trigger_id, event))
+        finally:
+            # CancelledError will get injected when we're stopped - which is
+            # fine, the cleanup process will understand that, but we want to
+            # allow triggers a chance to cleanup, either in that case or if
+            # they exit cleanly.
+            trigger.cleanup()
+
+    # Main-thread sync API
+
+    def update_triggers(self, requested_trigger_ids: Set[int]):
+        """
+        Called from the main thread to request that we update what
+        triggers we're running.
+
+        Works out the differences - ones to add, and ones to remove - then
+        adds them to the deques so the subthread can actually mutate the running
+        trigger set.
+        """
+        current_trigger_ids = set(self.triggers.keys())
+        # Work out the two difference sets
+        new_trigger_ids = requested_trigger_ids.difference(current_trigger_ids)
+        old_trigger_ids = current_trigger_ids.difference(requested_trigger_ids)
+        # Bulk-fetch new trigger records
+        new_triggers = Trigger.bulk_fetch(new_trigger_ids)
+        # Add in new triggers
+        for new_id in new_trigger_ids:
+            # Check it didn't vanish in the meantime
+            if new_id not in new_triggers:
+                self.log.warning("Trigger ID %s disappeared before we could start it", new_id)
+                continue
+            # Resolve trigger record into an actual class instance
+            trigger_class = self.get_trigger_by_classpath(new_triggers[new_id].classpath)
+            self.to_create.append((new_id, trigger_class(**new_triggers[new_id].kwargs)))
+        # Remove old triggers
+        for old_id in old_trigger_ids:
+            self.to_delete.append(old_id)
+
+    def get_trigger_by_classpath(self, classpath: str) -> Type[BaseTrigger]:
+        """
+        Gets a trigger class by its classpath ("path.to.module.classname")
+
+        Uses a cache dictionary to speed up lookups after the first time.
+        """
+        if classpath not in self.trigger_cache:
+            module_name, class_name = classpath.rsplit(".", 1)
+            try:
+                module = importlib.import_module(module_name)
+            except ImportError:
+                raise ImportError(
+                    f"Cannot import trigger module {module_name} (from trigger classpath {classpath})"
+                )
+            try:
+                trigger_class = getattr(module, class_name)
+            except AttributeError:
+                raise ImportError(f"Cannot import trigger {class_name} from module {module_name}")
+            self.trigger_cache[classpath] = trigger_class
```

Comment:
`from airflow.utils.module_loading import import_string` and then

```suggestion
            self.trigger_cache[classpath] = import_string(classpath)
```

(The exceptions would be slightly different)

Reply:
Gosh, the `utils` package keeps being a treasure-trove of handy things.

Reply:
Orrrr a junkyard where no one can find anything cos that's just where everything gets dumped.

Potato potato.